### PR TITLE
(SERVER-2895) Symlink to old cadir when setting up with the new default

### PR DESCRIPTION
**Pull request by mwaggett**
This commit adds a method to symlink the new default cadir to the old default
cadir location when running the `setup` and `import` commands. This ensures
that, even on new installs, we can access the cadir from the old location
because there are still tools in the ecosystem that expect the old location.

### CODE REVIEW

File: lib/puppetserver/ca/utils/file_system.rb
Comment (mwaggett) [inline lib/puppetserver/ca/utils/file_system.rb:53]: lmk if there's a better shared place to put this, nothing seemed like an obviously good place

File: lib/puppetserver/ca/utils/file_system.rb
Comment (mwaggett) [inline lib/puppetserver/ca/utils/file_system.rb:55]: Should I use the same `running_as_root?` branching that Justin had in his PR? That might make testing possible.. But not sure if we really only wanna do this linking if we're using that^ specific default.

**Comment by mwaggett** [comment]
I did a manual test of `puppetserver ca setup` on a FOSS install and it created the cadir in the right place and symlinked to the old place.

File: lib/puppetserver/ca/utils/file_system.rb
Comment (justinstoller) [reply lib/puppetserver/ca/utils/file_system.rb:55]: imo, having running from source be as similar to running from packages is something we should shoot for.

File: lib/puppetserver/ca/utils/file_system.rb
Comment (justinstoller) [inline lib/puppetserver/ca/utils/file_system.rb:57]: Files won't and shouldn't exist during this.

File: lib/puppetserver/ca/utils/file_system.rb
Comment (mwaggett) [reply lib/puppetserver/ca/utils/file_system.rb:55]: I'm not sure what you mean. Are you saying I should or should not do the same branching as you did?

File: lib/puppetserver/ca/utils/file_system.rb
Comment (mwaggett) [reply lib/puppetserver/ca/utils/file_system.rb:57]: ah yeah, I can make this comment more accurate

File: lib/puppetserver/ca/utils/file_system.rb
Comment (justinstoller) [reply lib/puppetserver/ca/utils/file_system.rb:53]: I really want to make this just a forceful symlink function and if it really needs to hardcode the ca defaults then move it to the `local_certificate_authority` but as is matches pretty well with how this module is structured. It _might_ be time to refactor it (the local ca class didn't exist when we made the file_system module) but I'm okay waiting a bit before doing a refactor.

File: lib/puppetserver/ca/utils/file_system.rb
Comment (justinstoller) [reply lib/puppetserver/ca/utils/file_system.rb:55]: Yeah, I'd prefer we do the branching.

File: lib/puppetserver/ca/utils/file_system.rb
Comment (mwaggett) [reply lib/puppetserver/ca/utils/file_system.rb:53]: I could make this more generic, but I think we'd still need another new method wrapped around it because we need to check whether we're using the new default or not before trying to link. Do you think that's worthwhile?

File: lib/puppetserver/ca/utils/file_system.rb
Comment (Magisus) [reply lib/puppetserver/ca/utils/file_system.rb:57]: I think that maybe means we don't need the `rm` right? Since this function is called during initial setup, and will be using the new location?

Comment (mwaggett) [reply lib/puppetserver/ca/utils/file_system.rb:57]: yeah, perhaps.. but seeing how much fun we've had with symlinks in the past, it seems safer to delete it just in case.

Comment (Magisus) [reply lib/puppetserver/ca/utils/file_system.rb:57]: Heh fair enough. Seems alright I guess.

File: lib/puppetserver/ca/utils/file_system.rb
Comment (tvpartytonight) [inline lib/puppetserver/ca/utils/file_system.rb:61]: These two calls can be combined into a single one using the [force](https://ruby-doc.org/stdlib-2.5.0/libdoc/fileutils/rdoc/FileUtils.html#method-i-ln_s) option.

```suggestion
FileUtils.symlink(new_cadir, old_cadir, force: true)
```

Comment (mwaggett) [reply lib/puppetserver/ca/utils/file_system.rb:61]: I don't think that will do what we want. If the target is a directory, the symlink will be nested under it. So we'd end up with `$old_cadir/$new_cadir -> $new_cadir`. I just tried this out in irb and it seems consistent with [the docs](https://ruby-doc.org/stdlib-2.7.1/libdoc/fileutils/rdoc/FileUtils.html#method-c-ln_s), but let me know if I've missed something.

Comment (tvpartytonight) [reply lib/puppetserver/ca/utils/file_system.rb]: Oh blerg, I misread that doc, good thing you tested it!

File: spec/utils/ssl.rb
Comment (Magisus) [inline spec/utils/ssl.rb:152]: Soo I'm actually not sure how I feel about this. This will clean up anything I have in my normal dev dir as a side effect of running these tests, which I feel could be unexpected and possibly destructive. But if others think we should treat that dev dir as unreliable (i.e. anything could mess with it at any time), I'll defer to that.

Comment (Magisus) [reply spec/utils/ssl.rb:152]: We could potentially back up what's there and put it back at the end?

File: lib/puppetserver/ca/local_certificate_authority.rb
Comment (Magisus) [inline lib/puppetserver/ca/local_certificate_authority.rb:54]: Can we utilize these where Justin had similar logic?

Comment (Magisus) [reply lib/puppetserver/ca/local_certificate_authority.rb:54]: Like somehow consolidate with https://github.com/puppetlabs/puppetserver-ca-cli/blob/main/lib/puppetserver/ca/config/puppet.rb#L53-L74

Comment (mwaggett) [reply lib/puppetserver/ca/local_certificate_authority.rb:54]: yeah, I was thinking about that. Wasn't sure how much we wanted to keep the config stuff as its own thing? but I can update.

Comment (Magisus) [reply lib/puppetserver/ca/local_certificate_authority.rb:54]: Could possibly put these in with the filesystem stuff and call them both places? Not totally sure what the right separation of concerns is, but the logic is finicky and hardcoded and I'd rather that only be in one place.

Comment (justinstoller) [reply lib/puppetserver/ca/local_certificate_authority.rb:54]: fwiw, I think these methods are available on the "puppet" config instance here: https://github.com/puppetlabs/puppetserver-ca-cli/pull/65/files#diff-8a63d3dddb93d86b02aba1619d4d8e877d75584395f677c0dbbce246e40d0b81R58

Passing either the whole config instance (since the settings are also a member of it) or gather these in the caller would probably work?

Comment (mwaggett) [reply lib/puppetserver/ca/local_certificate_authority.rb:54]: ugh diff links on PRs like that never work for me

Comment (mwaggett) [reply lib/puppetserver/ca/local_certificate_authority.rb:54]: but I think I know where you're talking about

Comment (Magisus) [reply lib/puppetserver/ca/local_certificate_authority.rb:54]: They don't work for me either.

Comment (mwaggett) [reply lib/puppetserver/ca/local_certificate_authority.rb:54]: hmm those puppet config methods are _not_ available in the tests.. Does it make sense to create an instance just to access those methods?

Comment (mwaggett) [reply lib/puppetserver/ca/local_certificate_authority.rb:54]: okay I just updated Justin's code to use these methods.

File: spec/utils/ssl.rb
Comment (mwaggett) [reply spec/utils/ssl.rb:152]: @justinstoller @tvpartytonight Do y'all have opinions on this?

File: lib/puppetserver/ca/local_certificate_authority.rb
Comment (Magisus) [reply lib/puppetserver/ca/local_certificate_authority.rb:54]: How would you feel about defining these in https://github.com/puppetlabs/puppetserver-ca-cli/blob/main/lib/puppetserver/ca/utils/config.rb as static functions? That feels like a better home for them than in this class. I don't _really_ like requiring this file in the config class (a little bit worried about winding up in a circular dependency by accident).

Comment (mwaggett) [reply lib/puppetserver/ca/local_certificate_authority.rb:54]: sure

Comment (mwaggett) [reply lib/puppetserver/ca/local_certificate_authority.rb]: done

File: lib/puppetserver/ca/utils/config.rb
Comment (Magisus) [inline lib/puppetserver/ca/utils/config.rb:55]: Should add the require for this file, to avoid any order-dependent issues.

File: lib/puppetserver/ca/local_certificate_authority.rb
Comment (Magisus) [reply lib/puppetserver/ca/local_certificate_authority.rb]: Thanks, I think that feels more right.

File: spec/utils/ssl.rb
Comment (tvpartytonight) [reply spec/utils/ssl.rb]: I feel more ok about modifying the `puppetserver/ca` directory, and less ok about deleting the `puppet/ssl` directory. I think that it is quite reasonable for this repo to modify the contents of `puppetserver/ca`, but `puppet/ssl` seems more of an unexpected side effect.

Not sure if this applies here, but could the structure of this be changed to allow for the tests to modify different locations, while the actual production code makes that impossible? For instance, the actual [migrate](https://github.com/puppetlabs/puppetserver-ca-cli/blob/main/lib/puppetserver/ca/action/migrate.rb#L57) call in `run` only ever calls it with 1 argument, but the actual method [definition](https://github.com/puppetlabs/puppetserver-ca-cli/blob/main/lib/puppetserver/ca/action/migrate.rb#L66-L69) has a default that can be overridden, the sole purpose for [easier testing](https://github.com/puppetlabs/puppetserver-ca-cli/blob/main/spec/puppetserver/ca/action/migrate_spec.rb#L85-L92).

File: lib/puppetserver/ca/utils/config.rb
Comment (mwaggett) [reply lib/puppetserver/ca/utils/config.rb:55]: ahh yes, thanks. so many moving parts haha

**Review by Magisus — APPROVED** [review]

**Comment by Magisus** [comment]
Will merge after squash.

File: spec/utils/ssl.rb
Comment (mwaggett) [reply spec/utils/ssl.rb]: hmm, I'm not sure if it applies either, since the way I'm testing this doesn't call `symlink_to_old_cadir` directly. But I'll think about this. And we can do Maggie's backup suggestion if I can't come up with something better.

**Comment by mwaggett** [comment]
Did another manual test for good measure and it still looks good. 
Just gonna think on the testing stuff over night, probs.

**Comment by justinstoller** [comment]
I think we should be setting the "default_{new,old}_cadir" to be relative to $confdir. That would let us test this easier and be consistent with the work we just landed.

**Comment by mwaggett** [comment]
@justinstoller relative to the actual confdir, as opposed to the defaults? I agree that would be easier to test, but then my question is: if someone is using a non-default confdir, do we still want to create the symlink? My understanding was that the point of this symlink is so tools that hard-code the (default) cadir will continue to work. If we symlink to a cadir under a non-default confdir, won't those tools still be broken?

**Comment by Magisus** [comment]
Wouldn't those tools already be broken if someone was using a non-default confdir then? Seems like that's not a case that actually happens in production (i.e. the confdir is expected to be default in that environment), so we can safely parameterize it without risking real use cases, but also allow easier testing.

**Comment by justinstoller** [comment]
So my understanding of the use cases for this tool are that 99% of it is for bootstrapping a new install with 0 custom configuration. There exists the ability to do custom configuration and some advanced users may do so. I expect those that have changed any of the defaults to "know what they're doing" and not need a symlink. However, if someone has set a $confdir to a custom value, but are otherwise expecting $cadir to be rooted in the same place as usual our new move could be disruptive.

Basically, I don't think users who are setting custom values (anywhere) when they use this command _need_ this symlink. However, I also don't think it's _incorrect_ behavior to create the symlink in the case of a custom $confdir case. I do think it is _incorrect_ to link a custom $cadir to the old cadir.

**Review by Magisus — APPROVED** [review]

File: lib/puppetserver/ca/utils/config.rb
Comment (justinstoller) [inline lib/puppetserver/ca/utils/config.rb:37]: why not just make the default in the arg list be puppet_confdir?

Comment (mwaggett) [reply lib/puppetserver/ca/utils/config.rb:37]: yeeeah, I thought about that.. it seemed kinda weird to me to call methods in the argument list, but I can change it.

**Review by Magisus — APPROVED** [review]

**Pull request merged**
